src/number_theory/pell.lean

```diff
@@ -398,4 +427,274 @@ end solution₁
+lemma d_nonsquare {a : solution₁ d} (h : is_fundamental a) : ¬ is_square d :=
+d_nonsquare_of_one_lt_x h.1
+
+/-- If there is a fundamental solution, it is unique. -/
+lemma unique {a b : solution₁ d} (ha : is_fundamental a) (hb : is_fundamental b) : a = b :=
+begin
+  have hx := le_antisymm (ha.2.2 hb.1) (hb.2.2 ha.1),
+  refine solution₁.ext hx _,
+  have : d * a.y ^ 2 = d * b.y ^ 2 := by rw [a.prop_y, b.prop_y, hx],
+  exact (sq_eq_sq ha.2.1.le hb.2.1.le).mp (int.eq_of_mul_eq_mul_left ha.d_pos.ne' this),
+end
+
+/-- If `d` is positive and not a square, then a fundamental solution exists. -/
+lemma exists_of_not_is_square (h₀ : 0 < d) (hd : ¬ is_square d) :
+  ∃ a : solution₁ d, is_fundamental a :=
+begin
+  obtain ⟨a, ha₁, ha₂⟩ := exists_pos_of_not_is_square h₀ hd,
+  -- convert to `x : ℕ` to be able to use `nat.find`
+  have P : ∃ x' : ℕ, 1 < x' ∧ ∃ y' : ℤ, 0 < y' ∧ (x' : ℤ) ^ 2 - d * y' ^ 2 = 1,
+  { have hax := a.prop,
+    lift a.x to ℕ using (by positivity) with ax,
+    norm_cast at ha₁,
+    exact ⟨ax, ha₁, a.y, ha₂, hax⟩, },
+  classical, -- to avoid having to show that the predicate is decidable
+  let x₁ := nat.find P,
+  obtain ⟨hx, y₁, hy₀, hy₁⟩ := nat.find_spec P,
+  refine ⟨mk x₁ y₁ hy₁, (by {rw x_mk, exact_mod_cast hx}), hy₀, λ b hb, _⟩,
+  rw x_mk,
+  have hb' := (int.to_nat_of_nonneg $ zero_le_one.trans hb.le).symm,
+  have hb'' := hb,
+  rw hb' at hb ⊢,
+  norm_cast at hb ⊢,
+  refine nat.find_min' P ⟨hb, |b.y|, abs_pos.mpr $ y_ne_zero_of_one_lt_x hb'', _⟩,
+  convert b.prop using 1,
+  rw [← hb', sq_abs],
+end
+
+/-- The `n`th power of a fundamental solution is trivial if and only if `n = 0`. -/
+lemma pow_eq_one_iff {a : solution₁ d} (h : is_fundamental a) (n : ℤ) : a ^ n = 1 ↔ n = 0 :=
+begin
+  refine ⟨λ H, _, λ H, by rw [H, zpow_zero]⟩,
+  have H' : (a ^ n).y = 0 := by rw [H, y_one],
+  rcases lt_trichotomy 0 n with hn | rfl | hn,
+  { exact absurd H' (y_zpow_pos h.x_pos h.2.1 hn).ne', },
+  { refl, },
+  { have := y_zpow_pos h.x_pos h.2.1 (by rwa [lt_neg, neg_zero] : 0 < -n),
+    rw [zpow_neg, y_inv, H', neg_zero] at this,
+    exact false.elim (lt_irrefl _ this), }
+end
+
+/-- The `n`th power of a fundamental solution has positive `y` if and only if `n` is positive. -/
+lemma pow_y_pos_iff {a : solution₁ d} (h : is_fundamental a) (n : ℤ) : 0 < (a ^ n).y ↔ 0 < n :=
+begin
+  refine ⟨λ H, _, y_zpow_pos h.x_pos h.2.1⟩,
+  rcases lt_trichotomy 0 n with hn | rfl | hn,
+  { exact hn, },
+  { rwa [zpow_zero, y_one] at H, },
+  { have := y_zpow_pos h.x_pos h.2.1 (by rwa [lt_neg, neg_zero] : 0 < -n),
+    rw [zpow_neg, y_inv, lt_neg, neg_zero] at this,
+    exact false.elim (lt_irrefl _ $ H.trans this), }
+end
+
+/-- The `n`th power of a fundamental solution has negative `y` if and only if `n` is negative. -/
+lemma pow_y_neg_iff {a : solution₁ d} (h : is_fundamental a) (n : ℤ) : (a ^ n).y < 0 ↔ n < 0 :=
+begin
+  rw [← neg_neg n, zpow_neg, y_inv, neg_lt, neg_zero, neg_lt, neg_zero],
+  exact h.pow_y_pos_iff (-n),
+end
```

Comment:
The last three lemmas would all follow from the statement that if a is a solution_1 and a.x, a.y>0 then the function from int to int sending n to (a^n).y is strictly monotonic, which might itself be easier to prove (one straightforward proof would be to show (a^(n+1)).y > (a^n).y and then claim that this is enough; the proof is obvious if n>=0 and is a sign argument if n<0). At least this would save you from having to apply `lt_trichotomy` more than once. Could you imagine that this would be a more useful approach or do you not think it's worth the hassle?

Reply:
I'll look into that. Is applying `lt_trichotomy` more than once a bad thing in and of itself?